### ISSUE TITLE
mat: Shadowing detection fails for embedded Vector

### ISSUE THREAD

**Issue by btracey** [gonum/gonum#945]
Found on https://stackoverflow.com/questions/51950841/gonum-throws-bad-region-panic-when-using-an-embedded-struct

Code:

```
package main

import (
	"fmt"

	"gonum.org/v1/gonum/mat"
)

type Vector struct {
	mat.VecDense
}

func NewVector(n int, data []float64) *Vector {
	return &Vector{*mat.NewVecDense(n, data)}
}

func main() {
	u, v := mat.NewVecDense(3, []float64{1, 2, 3}), mat.NewVecDense(3, []float64{4, 5, 6})
	fmt.Printf("[U - NewVecDense]\tADDRESS: %v, VALUE: %v\n", &u, u)
	fmt.Printf("[V - NewVecDense]\tADDRESS: %v, VALUE: %v\n", &v, v)
	rvu, rvv := u.RawVector(), v.RawVector()
	fmt.Printf("rvu %p, rvv %p\n", &rvu.Data[0], &rvv.Data[0])
	u.AddVec(u, v)

	fmt.Println("-------------------------")
	x, y := NewVector(3, []float64{1, 2, 3}), NewVector(3, []float64{4, 5, 6})
	fmt.Printf("[X - NewVector]\tADDRESS: %v, VALUE: %v\n", &x, x)
	rvx, rvy := x.RawVector(), y.RawVector()
	fmt.Printf("rvx %p, rvy %p\n", &rvx.Data[0], &rvy.Data[0])
	fmt.Printf("[Y - NewVector]\tADDRESS: %v, VALUE: %v\n", &y, y)
	x.AddVec(x, y)
	fmt.Println(x)
}
```

For some reason the second call fails while the first does not. Looking into it.

**Comment by btracey** [comment]
Oh, I see, it's because we check if the pointers of the two `VecDense` are equal. In the second case the pointers are not equal because the VecDense gets copied when it is passed into the function call. The underlying data slices are identical, even though the pointer is not. It is roughly equivalent to:

```
ap := mat.NewVecDense(3, nil)
a := *ap
b := a
a.AddVec(&a, &b)
```
Which similarly panics. I think this is a bug on our end, the above code should be fine to add. We should be checking if the data is the same, not the vector pointer value.

**Comment by kortschak** [comment]
The case you have here and the case in the SO post are different. Note that I [posted a comment there](https://stackoverflow.com/questions/51950841/gonum-throws-bad-region-panic-when-using-an-embedded-struct#comment93232664_51950841).

Solving the OP's problem would require that we use reflect, which we don't want to do.

I also don't think we want to make the case you have work; we provide a special case the `a.Op(a, b)` or `a.Op(b, a)` works. If we get more clever than that it gets more expensive and harder to document.

**Comment by btracey** [comment]
You're right, what I posted isn't the same. We do want to prevent the following case:
```
b := mat.NewVecDense(3, nil)
ap := mat.NewVecDense(3, nil)
a := *ap
c := a
c.AddVec(&a, b)
```
because that's exactly the point of shadowing detection. I agree that the case I posted above is also working as intended, `b` silently changes when `a` changes, and again, that's the point of shadowing detection.

OPs problem is tricky. I didn't see your comment until now. Your reply is "don't do it that way", but it's not clear to me this is a bad idea in general. One might wish to extend VecDense with methods `mat` doesn't provide, but then be able to use the built-in ones as well. Is there a way to do this I'm not seeing?

**Comment by kortschak** [comment]
There are two ways to do this.

1. Hide the `RawVector` (or other `Raw*`) method.
2. Provide all the methods and do the operations on the embedded values.

**Comment by matteo-grella** [comment]
@kortschak I'm having the same problem. Could you please elaborate a bit more with respect to your suggested way 1 (Hide the `RawVector`)?

**Comment by kortschak** [comment]
The shadow detection is only performed on types that satisfy the relevant `Raw*` interface. In your case this is `RawVectorer`. So what you should do is wrap the type we provide in a struct like so:
```
type myUnsafeType struct {
    Vector *mat.VecDense
}

func (v myUnsafeType) Dims() (r, c int) { return v.Vector.Dims() }
func (v myUnsafeType) At(r, c int) float64 { return v.Vector.At() }
func (v myUnsafeType) T() mat.Matrix { return v.Vector.T() }

etc
```

If you are going to be using a transpose, you'll also need to provide that and hook it in, otherwise the underlying type will escape the jail.

However, you probably don't need to do all this if you don't treat the language as an inheritance-based one. My favoured answer is still what I wrote [here](https://stackoverflow.com/questions/51950841/gonum-throws-bad-region-panic-when-using-an-embedded-struct#comment93232664_51950841).

**Comment by matteo-grella** [comment]
Thanks, I thought it was clear to me, but apparently not that much.
(Sorry I am quite new to Go!)

Let's say we have something like:

```
type UnsafeDense struct {
	*mat.Dense
}

func NewUnsafeDense(r, c int, data []float64) *Dense {
	return &UnsafeDense{mat.NewDense(r, c, data)}
}

func (m *UnsafeDense) Add(a, b mat.Matrix) {
	m.Dense.Add(a, b)
}
```

The `Add` will panic for the same reason (it doesn't matter if the receiver is a pointer or not).
What is the correct way to make it work?

**Comment by kortschak** [comment]
Yes. Don't embed. You'll note that in my code above I give the `*mat.VecDense` a name and do manual call dispatch to the underlying type. If you embed, you leave the `RawMatrix` method in the method set and the shadow detector will examine the elements. However, I don't see the reason behind needing to make a type that does this.

Methods in Go provide two major functions: interface satisfaction and namespacing (with the first being generally considered more important). If you don't need either of those, don't make new types with methods. This is the point to me comment at SO.

**Comment by matteo-grella** [comment]
> If you embed, you leave the RawMatrix method in the method set and the shadow detector will examine the elements.

Fully clear, thanks!

> However, I don't see the reason behind needing to make a type that does this.

Embedding, as well as using a delegate, allows us to enrich Gonum in our projects maintaining clean code. I'm dealing with `Dense`, `VecDense`,  [sparse.DOK](https://github.com/james-bowman/sparse) doing operations among them. 

Without method overloading and without using _a sort of_ inheritance, IMO it's a matter of seconds to have unmanageable names such as `DenseMulVecSparse`, `SparseMulVecSparseVec`.
